### ISSUE TITLE
Why has Colab stopped working for installing mediapipe-model-maker when it was fine just two days ago?

### ISSUE THREAD

**Issue by ferritti** [google-ai-edge/mediapipe#6070]
### Have I written custom code (as opposed to using a stock example script provided in MediaPipe)

None

### OS Platform and Distribution

macOS

### Python Version

3.12.11

### Task name (e.g. Image classification, Gesture recognition etc.)

Object Detection

### Describe the actual behavior

I had been following the "Object detection model customization guide". Everything ran successfully to the end of the code, and I even downloaded a separate file and trained it without issues. Up until two days ago it worked fine, but when I opened Colab today and tried again, I encountered an error right from the installation step.

### Describe the expected behaviour

I’m currently developing my thesis project and need to fine-tune my model, which makes solving this problem urgently necessary.

### Standalone code/steps you may have used to try to get what you need

https://colab.research.google.com/github/googlesamples/mediapipe/blob/main/examples/customization/object_detector.ipynb

### Other info / Complete Logs

```shell
Python 3.12.11
Requirement already satisfied: pip in /usr/local/lib/python3.12/dist-packages (24.1.2)
Collecting pip
  Downloading pip-25.2-py3-none-any.whl.metadata (4.7 kB)
Downloading pip-25.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 23.3 MB/s eta 0:00:00
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 24.1.2
    Uninstalling pip-24.1.2:
      Successfully uninstalled pip-24.1.2
Successfully installed pip-25.2
Collecting mediapipe-model-maker
  Downloading mediapipe_model_maker-0.2.1.4-py3-none-any.whl.metadata (1.7 kB)
Requirement already satisfied: absl-py in /usr/local/lib/python3.12/dist-packages (from mediapipe-model-maker) (1.4.0)
Collecting mediapipe>=0.10.0 (from mediapipe-model-maker)
  Downloading mediapipe-0.10.21-cp312-cp312-manylinux_2_28_x86_64.whl.metadata (9.7 kB)
Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from mediapipe-model-maker) (2.0.2)
Requirement already satisfied: opencv-python in /usr/local/lib/python3.12/dist-packages (from mediapipe-model-maker) (4.12.0.88)
INFO: pip is looking at multiple versions of mediapipe-model-maker to determine which version is compatible with other requirements. This could take a while.
Collecting mediapipe-model-maker
  Downloading mediapipe_model_maker-0.2.1.3-py3-none-any.whl.metadata (1.6 kB)
Requirement already satisfied: tensorflow>=2.10 in /usr/local/lib/python3.12/dist-packages (from mediapipe-model-maker) (2.19.0)
  Downloading mediapipe_model_maker-0.2.1.2-py3-none-any.whl.metadata (1.6 kB)
  Downloading mediapipe_model_maker-0.2.1.1-py3-none-any.whl.metadata (1.6 kB)
  Downloading mediapipe_model_maker-0.2.1-py3-none-any.whl.metadata (1.6 kB)
  Downloading mediapipe_model_maker-0.2.0-py3-none-any.whl.metadata (1.5 kB)
Requirement already satisfied: tensorflow-datasets in /usr/local/lib/python3.12/dist-packages (from mediapipe-model-maker) (4.9.9)
Requirement already satisfied: tensorflow-hub in /usr/local/lib/python3.12/dist-packages (from mediapipe-model-maker) (0.16.1)
Collecting tf-models-official==2.11.6 (from mediapipe-model-maker)
  Downloading tf_models_official-2.11.6-py2.py3-none-any.whl.metadata (1.5 kB)
Requirement already satisfied: Cython in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (3.0.12)
Requirement already satisfied: Pillow in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (11.3.0)
Requirement already satisfied: gin-config in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (0.5.0)
Requirement already satisfied: google-api-python-client>=1.6.7 in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (2.179.0)
Requirement already satisfied: immutabledict in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (4.2.1)
Requirement already satisfied: kaggle>=1.3.9 in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (1.7.4.5)
Requirement already satisfied: matplotlib in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (3.10.0)
Requirement already satisfied: oauth2client in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (4.1.3)
Requirement already satisfied: opencv-python-headless in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (4.12.0.88)
Requirement already satisfied: pandas>=0.22.0 in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (2.2.2)
Requirement already satisfied: psutil>=5.4.3 in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (5.9.5)
Requirement already satisfied: py-cpuinfo>=3.3.0 in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (9.0.0)
Requirement already satisfied: pycocotools in /usr/local/lib/python3.12/dist-packages (from tf-models-official==2.11.6->mediapipe-model-maker) (2.0.10)
Collecting pyyaml<6.0,>=5.1 (from tf-models-official==2.11.6->mediapipe-model-maker)
  Downloading PyYAML-5.4.1.tar.gz (175 kB)
  Installing build dependencies ... done
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Getting requirements to build wheel ... error
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Comment by hacktronics** [comment]
I am also facing the same issue.
tried this:
`pip install "cython<3.0.0" && pip install --no-build-isolation pyyaml==5.4.1`
but then it starts another set of errors, fix one, another one comes.

**Comment by ferritti** [comment]
Yes, thanks @hacktronics, I had tried that but unfortunately new errors keep coming up as I go along, I really don’t know how to fix it...

**Comment by ferritti** [comment]
@kuaashish do you know if there’s a way to fix this? I really need it, thanks for your help.

**Comment by HarishPermal** [comment]
Hi @ferritti, I was able to get the notebook running successfully by changing the runtime.

Solution:
In the "change runtime type" settings, set the `Runtime Version` to `2025.07`. This will switch the environment to use Python 3.11.13, which resolves the issue.

Could you try this and let me know if it works for you?

**Comment by ferritti** [comment]
Hi @HarishPermal, first of all, thanks for your help. I can confirm that it works, however another problem arises: the runtime disconnects after about 5 hours from the start of the model training. I’ve tried several times, but it keeps disconnecting, without giving the possibility to resume execution from where it stopped. It’s strange because in the free version of Colab you usually get up to 12 hours of execution time — in fact, before the update to the new Python version, I had reached up to 11 hours. In my case, I need more time…

I also tried using code snippets in the browser console to keep Colab active, but it was useless — apparently that’s not the problem.

**Comment by HarishPermal** [comment]
Hi @ferritti glad to hear the runtime version fix worked for you. 

Now regarding the disconnection after ~5hrs, a few workarounds you could try.

1. Using **Colab Pro/Pro+** they usually provide longer runtime
2. **Check pointing your model training**  like save weights/checkpoints every 30-60 mins to **Google Drive** so you can resume from where it left off.

Since the original problem **(Colab stopped working for installing mediapipe-model-maker when it was fine just two days ago?)** is resolved and the disconnection is due to **colab's free tier limits** not an issue with **MediaPipe**.

It would be great!  if you could go ahead and close this issue. but feel free 😊 to open a new one if you hit any MediaPipe-Specific errors.

**Comment by k0ndzi** [comment]
> Hi [@ferritti](https://github.com/ferritti), I was able to get the notebook running successfully by changing the runtime.
> 
> Solution: In the "change runtime type" settings, set the `Runtime Version` to `2025.07`. This will switch the environment to use Python 3.11.13, which resolves the issue.
> 
> Could you try this and let me know if it works for you?

it doesn’t work for me either. btw, how do you manage to change the runtime version to 2025.07? i don’t have such an option.

**Comment by HarishPermal** [comment]
Hi @k0ndzi ,

If you don't see the `Runtime version` option in Colab yet, you can still switch to `Python 3.11.13` which made [this](https://colab.research.google.com/github/googlesamples/mediapipe/blob/main/examples/customization/object_detector.ipynb) notebook run successfully in Colab by manually running the following in Colab cell

```bash
!sudo apt-get update -y
!sudo apt-get install -y python3.11 python3.11-distutils

!sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.11 1
!sudo update-alternatives --set python3 /usr/bin/python3.11

!curl -sS https://bootstrap.pypa.io/get-pip.py | python3

!python3 --version
!pip --version
```
if the above code runs successfully you should see

```print
Python 3.11.13
pip 25.2 from /usr/local/lib/python3.11/dist-packages/pip (python 3.11)
```

This will give you **Python 3.11.13** even if the runtime version drop-down isn't available on your account yet.

Hope this helps 😉, and do let me know if it resolves the issue on your side.

**Comment by ferritti** [comment]
Hi @HarishPermal, I tried with the Colab PRO version, but after a few hours the Runtime got disconnected, causing me to lose my progress. I believe the previous Runtime version is unstable right now, since even with a subscription the result doesn’t change. At this point, I don’t know what else to do—my last hope is to use the checkpoints you recommended. Could you explain what I need to add to the code to use these checkpoints, and how to resume from the last save if the Runtime disconnects again?

**Comment by HarishPermal** [comment]
Hi @ferritti ,

Thanks for your patience with the issue, To solve this, we had to manually reconstruct the model creation, compilation and training steps.

replace the existing model creation
```python
model = object_detector.ObjectDetector.create(
    train_data=train_data,
    validation_data=validation_data,
    options=options)
```

with the below, And replace the `export_dir='/content/drive/MyDrive/mediapipe_issue/issue_6070/exported_model'` with your Google drive path.
```python3
from mediapipe_model_maker.python.vision.object_detector import model_spec
from mediapipe_model_maker.python.vision.object_detector import model_options

spec_enum = object_detector.SupportedModels.MOBILENET_MULTI_AVG
spec = model_spec.SupportedModels.get(spec_enum)
hparams = object_detector.HParams(export_dir='/content/drive/MyDrive/mediapipe_issue/issue_6070/exported_model', epochs=20)
options = object_detector.ObjectDetectorOptions(supported_model=spec_enum, hparams=hparams)

current_model_options = options.model_options if options.model_options else model_options.ObjectDetectorModelOptions()

checkpoint_dir = os.path.join(hparams.export_dir, 'checkpoints')
os.makedirs(checkpoint_dir, exist_ok=True)

print("Initializing model...")
model = object_detector.ObjectDetector(
    model_spec=spec,
    label_names=train_data.label_names,
    hparams=hparams,
    model_options=current_model_options
)

print("Creating and compiling model...")
steps_per_epoch = train_data.size // hparams.batch_size
optimizer = model._create_optimizer(steps_per_epoch=steps_per_epoch)
model._create_model()
model._model.compile(optimizer=optimizer)

initial_epoch = 0 
latest_checkpoint = tf.train.latest_checkpoint(checkpoint_dir)
if latest_checkpoint:
    print(f"Resuming training from checkpoint: {latest_checkpoint}")
    model._model.load_weights(latest_checkpoint)
    initial_epoch = int(latest_checkpoint.split('-')[-1])
else:
    print("Starting training from scratch.")


checkpoint_callback = tf.keras.callbacks.ModelCheckpoint(
    filepath=os.path.join(checkpoint_dir, 'ckpt-{epoch}'),
    save_weights_only=True,
    verbose=1,
    save_freq='epoch'
)

print("Preparing datasets for training...")
train_dataset = train_data.gen_tf_dataset(
    batch_size=hparams.batch_size,
    is_training=True,
    shuffle=hparams.shuffle,
    preprocess=model._preprocessor
).repeat()

validation_dataset = validation_data.gen_tf_dataset(
    batch_size=hparams.batch_size,
    is_training=False,
    preprocess=model._preprocessor
)

print(f"Starting model training for {hparams.epochs} epochs...")
history = model._model.fit(
    x=train_dataset,
    epochs=hparams.epochs,
    initial_epoch=initial_epoch, 
    steps_per_epoch=steps_per_epoch if steps_per_epoch > 0 else None,
    validation_data=validation_dataset,
    callbacks=[checkpoint_callback]
)

print("Training complete. Saving final float checkpoint.")
model._save_float_ckpt()
```
Please try this and let me know 🙃.

**Comment by ferritti** [comment]
Hi @HarishPermal, I confirm that the issue has been resolved thanks to the checkpoints. The runtime disconnects after a few hours, but it is then possible to resume training from the last epoch and export the model once it is finished. I would like to thank you for your patience and the time you dedicated to my problem. I will go ahead and close the issue.

**Comment by sustinespacne9-ai** [comment]
> Hi [@ferritti](https://github.com/ferritti) ,
> 
> Thanks for your patience with the issue, To solve this, we had to manually reconstruct the model creation, compilation and training steps.
> 
> replace the existing model creation
> 
> model = object_detector.ObjectDetector.create(
>     train_data=train_data,
>     validation_data=validation_data,
>     options=options)
> with the below, And replace the `export_dir='/content/drive/MyDrive/mediapipe_issue/issue_6070/exported_model'` with your Google drive path.
> 
> from mediapipe_model_maker.python.vision.object_detector import model_spec
> from mediapipe_model_maker.python.vision.object_detector import model_options
> 
> spec_enum = object_detector.SupportedModels.MOBILENET_MULTI_AVG
> spec = model_spec.SupportedModels.get(spec_enum)
> hparams = object_detector.HParams(export_dir='/content/drive/MyDrive/mediapipe_issue/issue_6070/exported_model', epochs=20)
> options = object_detector.ObjectDetectorOptions(supported_model=spec_enum, hparams=hparams)
> 
> current_model_options = options.model_options if options.model_options else model_options.ObjectDetectorModelOptions()
> 
> checkpoint_dir = os.path.join(hparams.export_dir, 'checkpoints')
> os.makedirs(checkpoint_dir, exist_ok=True)
> 
> print("Initializing model...")
> model = object_detector.ObjectDetector(
>     model_spec=spec,
>     label_names=train_data.label_names,
>     hparams=hparams,
>     model_options=current_model_options
> )
> 
> print("Creating and compiling model...")
> steps_per_epoch = train_data.size // hparams.batch_size
> optimizer = model._create_optimizer(steps_per_epoch=steps_per_epoch)
> model._create_model()
> model._model.compile(optimizer=optimizer)
> 
> initial_epoch = 0 
> latest_checkpoint = tf.train.latest_checkpoint(checkpoint_dir)
> if latest_checkpoint:
>     print(f"Resuming training from checkpoint: {latest_checkpoint}")
>     model._model.load_weights(latest_checkpoint)
>     initial_epoch = int(latest_checkpoint.split('-')[-1])
> else:
>     print("Starting training from scratch.")
> 
> 
> checkpoint_callback = tf.keras.callbacks.ModelCheckpoint(
>     filepath=os.path.join(checkpoint_dir, 'ckpt-{epoch}'),
>     save_weights_only=True,
>     verbose=1,
>     save_freq='epoch'
> )
> 
> print("Preparing datasets for training...")
> train_dataset = train_data.gen_tf_dataset(
>     batch_size=hparams.batch_size,
>     is_training=True,
>     shuffle=hparams.shuffle,
>     preprocess=model._preprocessor
> ).repeat()
> 
> validation_dataset = validation_data.gen_tf_dataset(
>     batch_size=hparams.batch_size,
>     is_training=False,
>     preprocess=model._preprocessor
> )
> 
> print(f"Starting model training for {hparams.epochs} epochs...")
> history = model._model.fit(
>     x=train_dataset,
>     epochs=hparams.epochs,
>     initial_epoch=initial_epoch, 
>     steps_per_epoch=steps_per_epoch if steps_per_epoch > 0 else None,
>     validation_data=validation_dataset,
>     callbacks=[checkpoint_callback]
> )
> 
> print("Training complete. Saving final float checkpoint.")
> model._save_float_ckpt()
> Please try this and let me know 🙃.

Thank you for the detailed workaround, I really appreciate it.
However, the issue I'm encountering seems to happen earlier during the hyperparameter initialization, where the model still attempts to access the TensorFlow Model Garden GCS bucket for the pretrained checkpoint. The error I’m getting is related to the GCS URL returning 403 (forbidden).
From what I understand, the problem is not only in the model creation or training steps, but in the checkpoint download step itself. It seems the hyperparameters or model spec still reference the original GCS path internally.
In the previous suggestion, the workaround involved manually downloading the checkpoint from GCS and forcing training to start locally. I’m trying to understand how to fully override the default GCS dependency during model spec initialization.
Could you clarify whether your solution completely bypasses the GCS download step, or if it still attempts to access it internally?
